### PR TITLE
Dsl Builder

### PR DESCRIPTION
**Pull request by smallshen**
https://github.com/JetBrains/KotlinDL/pull/252

@zaleslaw

### CODE REVIEW

**Comment by zaleslaw** [comment]
Thank you so much for separate PR

**Comment by juliabeliaeva** [comment]
With this dsl we can declare layer variables like this (not possible with `Functional.of`):
```kotlin 
functional {
  layers {
    val input = +Input(128, 128)
    +Conv1D()(input)
  }
}
```

**Comment by smallshen** [comment]
@juliabeliaeva 
‘’’kotlin
functional {
  layers {
    +Conv1D()(+Input(128, 128))
  }
}
‘’’

this could be possible too

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDsl.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    public fun layers(builder: LayerBuilder.() -> Unit) {
+        this.layerProvider = builder
+    }
+
+    @BuilderDsl
+    public fun use(builder: T.() -> Unit) {
+        use = builder
+    }
+
+    public fun build(): T = initializer.invoke(LayerBuilder().apply(layerProvider).toArray()).apply(handler).apply {
```
Comment (juliabeliaeva) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDsl.kt:37]: This line is too long and hard to read, could be reformated into several lines.

**Review by juliabeliaeva — CHANGES_REQUESTED** [review]
This PR looks great, thank you. I'd add kdocs and improve tests first, naming comments are more of a cosmetics.

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDsl.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+@DslMarker
+@Retention(AnnotationRetention.SOURCE)
+internal annotation class BuilderDsl
+
+@DslMarker
+@Retention(AnnotationRetention.SOURCE)
+internal annotation class EntryDsl
+
+public class GraphTrainableModelDsl<T : GraphTrainableModel>(private val initializer: (Array<Layer>) -> T) {
```
[the remark this answers is not in the document]
Comment (smallshen) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDsl.kt]: renamed class GraphTrainableModelDsl to GraphTrainableModelBuilder, LayerBuilder to LayerListBuilder

[the remark this answers is not in the document]
Comment (smallshen) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDsl.kt]: Add documentation of unaryPlus

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDsl.kt [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    private var handler: T.() -> Unit = {}
+    private var layerProvider: LayerBuilder.() -> Unit = {}
+    private var use: ((T) -> Unit)? = null
```
[the remark this answers is not in the document]
Comment (smallshen) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDsl.kt]: rename `handler` to `modelBuilderBlock`, `layerProvider` to `layerListBuilderBlock` and  `use` to `useBlock`

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDsl.kt [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    @BuilderDsl
+    public fun model(builder: T.() -> Unit) {
+        this.handler = builder
```
[the remark this answers is not in the document]
Comment (smallshen) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDsl.kt]: Removed

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDsl.kt [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    @BuilderDsl
+    public fun layers(builder: LayerBuilder.() -> Unit) {
+        this.layerProvider = builder
+    }
+
+    @BuilderDsl
+    public fun use(builder: T.() -> Unit) {
```
[the remark this answers is not in the document]
Comment (smallshen) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDsl.kt]: rename it to block

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDsl.kt (hunk 1 above)
Comment (smallshen) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDsl.kt]: renamed initializer createModelFunction and use invoke operator function.

File: api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDslTest.kt [hunk 6]
```diff
@@ -0,0 +1,54 @@
+package org.jetbrains.kotlinx.dl.api.core.dsl
+
+import org.jetbrains.kotlinx.dl.api.core.Functional
+import org.jetbrains.kotlinx.dl.api.core.layer.convolutional.Conv1D
+import org.jetbrains.kotlinx.dl.api.core.layer.core.Input
+import org.junit.jupiter.api.Assertions.*
+import org.junit.jupiter.api.Test
+
+internal class GraphTrainableModelDslTest {
```
[the remark this answers is not in the document]
Comment (smallshen) [reply api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDslTest.kt]: It is used to check that the DSL does not throw an exception. For comparison, maybe we can compare the layers in GraphTrainableModel::layers.

File: api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelBuilderTest.kt [hunk 7]
```diff
@@ -19,36 +20,41 @@ internal class GraphTrainableModelBuilderTest {
...
-        assertDoesNotThrow {
+        assert(
             functional {
                 model {
                     name = "Test Functional"
                 }
 
                 layers {
                     +Input(128, 128)
-                    +Conv1D()
```
Comment (smallshen) [inline api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelBuilderTest.kt]: https://github.com/JetBrains/KotlinDL/issues/314
Temporary workaround to make test pass

**Comment by zaleslaw** [comment]
Thanks for the review @juliabeliaeva 

@smallshen 
From my side, I also suggest adding an example (or test) with the analog of building toyresnet in this 
```
functional {
    layers {
        +Conv1D()(+Input(128, 128))
    }
}
```
style.

What do you think, is it easy to extend in the future or combine with the existing Image Preprocessing DSL and non-existing yet DSL for training as the following:

```
val model = trainableModel {
      preprocess {
            load {
                pathToData = ...
            }
            transformImage {
                resize {
                    outputWidth = 224
                    outputHeight = 224
                }
                convert { colorMode = ColorMode.BGR }
            }
        }
  
  functional {
      layers {
          +Conv1D()(+Input(128, 128))
          ...
      }
   }
  compile {
        optimizer = Adam()
        loss = CrossEntropy()
        metrics = Accuracy()
  }
}
```

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDsl.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (juliabeliaeva) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/dsl/GraphTrainableModelDsl.kt]: Please also add docs for `GraphTrainableModelBuilder`.

**Review by juliabeliaeva — APPROVED** [review]

**Comment by zaleslaw** [comment]
Thanks for the PR, @smallshen hope it will be an important part of the future KotlinDL DSL family

**Pull request merged**